testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs

```diff
@@ -0,0 +1,297 @@
+        let client = APIClient::new(config);
+        let validator_to_node = Arc::new(Mutex::new(HashMap::new()));
+        let fullnode_to_node = Arc::new(Mutex::new(HashMap::new()));
+        Ok(Self {
+            client,
+            validator_to_node,
+            fullnode_to_node,
+        })
+    }
+
+    fn validator_spec(
+        &self,
+        index: u32,
+        num_validators: u32,
+        num_fullnodes: u32,
+        node_name: &str,
+        image_tag: &str,
+        delete_data: bool,
+    ) -> Vec<u8> {
+        let pod_yaml = format!(
+            include_str!("validator_spec_template.yaml"),
+            index = index,
+            num_validators = num_validators,
+            num_fullnodes = num_fullnodes,
+            image_tag = image_tag,
+            node_name = node_name,
+            cfg_overrides = "",
+            delete_data = delete_data,
+        );
+        let pod_spec: serde_yaml::Value = serde_yaml::from_str(&pod_yaml).unwrap();
+        serde_json::to_vec(&pod_spec).unwrap()
+    }
+
+    fn fullnode_spec(
+        &self,
+        fullnode_index: u32,
+        num_fullnodes: u32,
+        validator_index: u32,
+        num_validators: u32,
+        node_name: &str,
+        image_tag: &str,
+        delete_data: bool,
+    ) -> Vec<u8> {
+        let pod_yaml = format!(
+            include_str!("fullnode_spec_template.yaml"),
+            fullnode_index = fullnode_index,
+            num_fullnodes = num_fullnodes,
+            validator_index = validator_index,
+            num_validators = num_validators,
+            node_name = node_name,
+            image_tag = image_tag,
+            cfg_overrides = "",
+            delete_data = delete_data,
+        );
+        let pod_spec: serde_yaml::Value = serde_yaml::from_str(&pod_yaml).unwrap();
+        serde_json::to_vec(&pod_spec).unwrap()
+    }
+
+    async fn get_pod_node_and_ip(&self, pod_name: &str) -> Result<(String, String)> {
+        retry::retry_async(retry::fixed_retry_strategy(5000, 60), || {
+            let pod_api = Api::v1Pod(self.client.clone()).within(DEFAULT_NAMESPACE);
+            let pod_name = pod_name.to_string();
+            Box::pin(async move {
+                match pod_api.get(&pod_name).await {
+                    Ok(o) => {
+                        let node_name = o.spec.node_name.ok_or_else(|| {
+                            format_err!("node_name not found for pod {}", pod_name)
+                        })?;
+                        let pod_ip = o
+                            .status
+                            .ok_or_else(|| format_err!("status not found for pod {}", pod_name))?
+                            .pod_ip
+                            .ok_or_else(|| format_err!("pod_ip not found for pod {}", pod_name))?;
+                        if node_name.is_empty() || pod_ip.is_empty() {
+                            bail!("node_name not found for pod {}", pod_name)
```

Comment:
Is there legitimate case when it can be empty String (and not None)? If so, should we have different error here to distinguish them?